shared/oracle/tx_price_oracle.go

```diff
@@ -0,0 +1,102 @@
+package oracle
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/skip-mev/go-fast-solver/shared/clients/coingecko"
+	"github.com/skip-mev/go-fast-solver/shared/config"
+)
+
+const (
+	coingeckoUSDCurrency = "usd"
+)
+
+type TxPriceOracle interface {
+	// TxFeeUUSDC estimates what the cost in uusdc would be to execute a tx. The
+	// tx's gas fee cap and gas limit must be set.
+	TxFeeUUSDC(ctx context.Context, tx *types.Transaction) (*big.Int, error)
+
+	// GasCostUUSDC converts a tx fee to uusdc based on the current CoinGecko of
+	// the gas token in usd.
+	GasCostUUSDC(ctx context.Context, txFee *big.Int, chainID string) (*big.Int, error)
+}
+
+// Oracle is a evm uusdc tx execution price oracle that determines the price of
+// executing a tx on chain in uusdc.
+type Oracle struct {
+	coingecko coingecko.PriceClient
+}
+
+// NewOracle creates a new evm uusdc tx execution price oracle.
+func NewOracle(coingecko coingecko.PriceClient) *Oracle {
+	return &Oracle{coingecko: coingecko}
+}
+
+// TxFeeUUSDC estimates what the cost in uusdc would be to execute a tx. The
+// tx's gas fee cap and gas limit must be set.
+func (o *Oracle) TxFeeUUSDC(ctx context.Context, tx *types.Transaction) (*big.Int, error) {
+	if tx.Type() != types.DynamicFeeTxType {
+		return nil, fmt.Errorf("tx type must be dynamic fee tx, got %d", tx.Type())
+	}
+
+	// for a dry ran tx, GasFeeCap() will be the suggested gas tip cap + base
+	// fee of current chain head
+	estimatedPricePerGas := tx.GasFeeCap()
+	if estimatedPricePerGas == nil {
+		return nil, fmt.Errorf("tx's gas fee cap must be set")
+	}
+
+	// for a dry ran tx, Gas() will be the result of calling eth_estimateGas
+	estimatedGasUsed := tx.Gas()
+
+	txFee := new(big.Int).Mul(estimatedPricePerGas, big.NewInt(int64(estimatedGasUsed)))
+	return o.GasCostUUSDC(ctx, txFee, tx.ChainId().String())
+}
+
+// GasCostUUSDC converts a tx fee to uusdc based on the current CoinGecko of
+// the gas token in usd.
+func (o *Oracle) GasCostUUSDC(ctx context.Context, txFee *big.Int, chainID string) (*big.Int, error) {
```

Comment:
made this public, no need for it to be private and it works for all chain types, not just evm

Reply:
cosmos chains can have multiple gas tokens but i dont think we need to worry about that here

Reply:
yeah i thought of that but my thought was that we are only going to be using osmosis for the foreseeable future and if we wanted to support another destination chain probably a lot of other things are going to have to change so this is probably a fine assumption to make of now